java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java

```diff
@@ -0,0 +1,150 @@
+    }
+  }
+
+  @Test
+  public void testParquetWriteWithPartitions() throws Exception {
+    ParquetWriteSupport writeSupport = ParquetWriteSupport.writeTempFile(AVRO_SCHEMA_USER, TMP.newFolder(),
+        1, "a", 2, "b", 3, "c", 2, "d");
+    String sampleParquet = writeSupport.getOutputURI();
+    ScanOptions options = new ScanOptions(new String[0], 100);
+    final File writtenFolder = TMP.newFolder();
+    final String writtenParquet = writtenFolder.toURI().toString();
+
+    try (FileSystemDatasetFactory factory = new FileSystemDatasetFactory(rootAllocator(), NativeMemoryPool.getDefault(),
+        FileFormat.PARQUET, sampleParquet);
+         final Dataset dataset = factory.finish();
+         final Scanner scanner = dataset.newScan(options);
+         final ArrowScannerReader reader = new ArrowScannerReader(scanner, rootAllocator());
+         ) {
+      DatasetFileWriter.write(rootAllocator(), reader, FileFormat.PARQUET, writtenParquet, new String[]{"id", "name"}, 100, "data_{i}");
+      final Set<String> expectedOutputFiles = new HashSet<>(
+          Arrays.asList("id=1/name=a/data_0", "id=2/name=b/data_0", "id=3/name=c/data_0", "id=2/name=d/data_0"));
+      final Set<String> outputFiles = FileUtils.listFiles(writtenFolder, null, true)
+          .stream()
+          .map(file -> {
+            return writtenFolder.toURI().relativize(file.toURI()).toString();
+          })
+          .collect(Collectors.toSet());
+      Assert.assertEquals(expectedOutputFiles, outputFiles);
+    }
+  }
+
+  private void assertParquetFileEquals(String expectedURI, String actualURI) throws Exception {
+    final FileSystemDatasetFactory expectedFactory = new FileSystemDatasetFactory(
+        rootAllocator(), NativeMemoryPool.getDefault(), FileFormat.PARQUET, expectedURI);
+    List<ArrowRecordBatch> expectedBatches = collectResultFromFactory(expectedFactory,
+        new ScanOptions(new String[0], 100));
+    final FileSystemDatasetFactory actualFactory = new FileSystemDatasetFactory(
+        rootAllocator(), NativeMemoryPool.getDefault(), FileFormat.PARQUET, actualURI);
+    List<ArrowRecordBatch> actualBatches = collectResultFromFactory(actualFactory,
+        new ScanOptions(new String[0], 100));
+    // fast-fail by comparing metadata
+    Assert.assertEquals(expectedBatches.toString(), actualBatches.toString());
+    // compare ArrowRecordBatches
+    VectorSchemaRoot expectVsr = VectorSchemaRoot.create(expectedFactory.inspect(), rootAllocator());
+    VectorLoader expectLoader = new VectorLoader(expectVsr);
+    for(ArrowRecordBatch batch: expectedBatches) {
+      expectLoader.load(batch);
+    }
+
+    VectorSchemaRoot actualVsr = VectorSchemaRoot.create(actualFactory.inspect(), rootAllocator());
+    VectorLoader actualLoader = new VectorLoader(actualVsr);
+    for(ArrowRecordBatch batch: actualBatches) {
+      actualLoader.load(batch);
```

Comment:
This needs to be interleaved with the comparison, else you are actually only comparing the last batch

Reply:
Update to compare all the batches.